### PR TITLE
New Options, Generalizations, Thoughts

### PR DESCRIPTION
**Pull request by jmcbroome**
This PR contains a bunch of new options I wrote to disable some of the more expensive/messy default behaviors and removes some assumptions about the user passing masking information and diploid data. It is in no way comprehensive for every change I might want to make, but it was able to accomplish my immediate goal of making a diff from a minimal matUtils produced VCF. I tested it with the following commands:

```
matUtils extract -i public-latest.all.masked.pb.gz -z 20 -v test.vcf
python3 parsevcf/vcf_to_diff_script.py -v test.vcf -ha -od test.diff
```

PLEASE NOTE: while all original behavior can be reproduced with the correct options, I did turn off some things by default- like producing masking reports for every sample. I also did not touch the WDL. The WDL commands will need to be updated accordingly to keep your behavior consistent if you merge this! You should also test this on your existing data just to make sure I didn't break anything!

Some things I changed:
1. Cleaned up arguments and helpstrings. Better logging level setting. 
2. Make reference length and name into options instead of hardcoded assumptions. Default is the TB genome name/length (so same as previous behavior)
3. Make the numerous masking steps optional. Set default behaviors for when no masking information is passed.
4. Make the bcftools call step optional, as well as the overall masking being optional. 
5. Alter output behavior. Writing masking reports and diffs for every single sample is now optional and off by default. The user can get a single concatenated diff with the new option -od.
6. Add (simple) haploid support and and associated option- haploids are treated as homozygous diploids. Please make sure this doesn't have any unintended consequences.
7. Adjusted handling of gzipped input/vcf input in general. Your users will have ".vcf.gz" at the end of their files if their files are gzipped, so just check for that. And if they don't, the user is in the wrong! Additionally, just checking whether its binary could lead to problems if they used other compression formats. Added some try/catch as well. 
8. Basic installation/usage to README
9. General docstring tweaks.

Some general thoughts:

Keeping giant commented-out blocks of code is never a good look- makes it hard to navigate. You can always find old code in your github history if you need it in the future. Similarly, just go ahead and enable the debug-level logging statements for the debug-level- and if you don't want them even there, then just delete them!

There's an implementation question re: disk versus memory use. Right now you split a VCF by writing it iteratively to disk. This gets around loading extremely large files into memory- valuable, but I/O bound and potentially messy. Loading it into memory, comparatively, is simpler and means your work will not be I/O bound, but means you need to have a lot of things in RAM. You may want to consider supporting either behavior- disk writing sample-level variation as an optional behavior for users with extremely large VCF. This is more of a long-term thing to consider than a requirement, though. 

Ideally, the final version of this will not call bcftools at any point. Other than that one step its dependency-free, and that will make maintaining, distributing, and using this tool much easier! If you could replicate the bcftools behavior in Python or just remove that step altogether, it would go a way towards making this more of an independent software tool.

Masking is a huge part of what this script is doing. I've done some super simple additions to the README, but a brief overview of the diff format and details on your masking parameters would be critical additions to the documentation. 

Let me know your thoughts. This is a generally impressive piece of work (appreciate the use of logging and docstrings), though in need of some TLC in terms of organization/clarity. If you'd like to chat about what this would need to get published on pip, let me know!

### CODE REVIEW

**Comment by aofarrel** [comment]
Adding a WDL perspective to this (caveat: I'm not an expert on hardware)...

> There's an implementation question re: disk versus memory use. Right now you split a VCF by writing it iteratively to disk. This gets around loading extremely large files into memory- valuable, but I/O bound and potentially messy. Loading it into memory, comparatively, is simpler and means your work will not be I/O bound, but means you need to have a lot of things in RAM. You may want to consider supporting either behavior- disk writing sample-level variation as an optional behavior for users with extremely large VCF. This is more of a long-term thing to consider than a requirement, though.

When running a WDL on Terra, the WDL asks Google for the following:
* How much disk space (treated as a hard limit -- if you go over this, the workflow dies)
* Whether that disk is an HDD or an SSD
* How much RAM 
* How many CPUs
* Some other stuff not relevant to this discussion

All of this gets requested before anything in your script is run. This means that the WDL script -- or more accurately, the guy writing the WDL script -- needs to be able to predict how much disk size it needs. What I do is write a formula that scales on the size of the inputs (which WDL knows before runtime in this context), for example:

`Int finalDiskSize = ceil(size(bam, "GB")) + 2*ceil(size(vcf, "GB")) + addldisk`

basically means "ask Google for the size of the bam file in GB rounded up the nearest integer, plus double the size of the vcf file in GB rounded up the nearest integer, plus addldisk." addldisk is an integer variable that the user can set that I use as bandaid for the fact that my formula might not be super accurate and the user can just keep throwing gigabytes at it until it works. For most of my WDLs I set addldisk to 10 GB. For this WDL I set it much higher because it's got a lot of I/O.

The thing is, when it comes to GCP, disk space is cheap, and memory is expensive. Also, for the TB diffs we've been making so far, your script runs very quickly. So I would argue that in regards to WDL, the current behavior works quite well even for large VCFs. However, I agree that writing to memory is an important option for the long term, especially for running this script outside of GCP.

TLDR -- I don't disagree with Jakob, all I ask is that memory-based compute doesn't completely replace I/O based compute. I think his proposal of having it be an option as a long-term plan would be great.

**Comment by jmcbroome** [comment]
That's quite interesting! I'm not at all familiar with GCP pricing or performance behavior and have no doubts that you're correct that the current implementation works well in that environment. I brought that point up mostly from the perspective of a research user who is working locally with a smaller dataset. To reiterate, it's an option I think might be useful in the long term, and is certainly neither an essential immediate upgrade, nor a good replacement for the current behavior.
